File: packages/core/src/providers/factory.ts
import { OpenAICompatibleProvider } from "./openai-compatible.js";
import type { OpenAIProviderOptions } from "./openai-compatible.js";
import { BedrockProvider } from "./bedrock.js";
import type { ProviderConfig, LLMProvider, ModelInfo } from "../types.js";
import { resolveApiKey } from "../config.js";
import { getCopilotToken } from "./github-auth.js";
import type { DeviceFlowCallbacks } from "./github-auth.js";

/** Pre-configured presets — just add model + optional overrides */
const PRESETS: Record<string, (config: ProviderConfig) => OpenAIProviderOptions> = {
  ollama: (c) => ({
    name: "ollama",
    model: c.model,
    baseUrl: c.baseUrl ?? "http://localhost:11434/v1",
    maxTokens: c.maxTokens,
  }),

  "github-copilot": (c) => ({
    name: "github-copilot",
    model: c.model,
    baseUrl: c.baseUrl ?? "https://api.githubcopilot.com",
    getApiKey: () => getCopilotToken(),
    headers: { "Copilot-Integration-Id": "vscode-chat" },
    maxTokens: c.maxTokens,
  }),

  openai: (c) => ({
    name: "openai",
    model: c.model,
    baseUrl: c.baseUrl ?? "https://api.openai.com/v1",
    // `resolveApiKey` covers the env-var and file forms, so a cluster
    // deployment never needs a literal key in its config file.
    apiKey: resolveApiKey(c) ?? process.env.OPENAI_API_KEY,
    maxTokens: c.maxTokens,
  }),
};

/** Create a provider from config — uses presets or falls back to raw OpenAI-compat */
export function createProvider(config: ProviderConfig): LLMProvider {
  if (config.type === "bedrock") {
    return new BedrockProvider({
      name: config.name ?? "bedrock",
      model: config.model,
      region: config.region,
      maxTokens: config.maxTokens,
      thinking: config.thinking,
      effort: config.effort,
    });
  }

  const preset = PRESETS[config.type];
  if (preset) {
    return new OpenAICompatibleProvider(preset(config));
  }
  // Fallback: treat type as a custom OpenAI-compatible endpoint
  return new OpenAICompatibleProvider({
    name: config.name ?? config.type,
    model: config.model,
    baseUrl: config.baseUrl ?? "",
    apiKey: resolveApiKey(config),
    maxTokens: config.maxTokens,
  });
}

/**
 * List available models for a provider type.
 * Creates a temporary provider instance just for the models query.
 * For github-copilot, this triggers the device auth flow if needed.
 */
export async function listModelsForProvider(
  type: string,
  opts?: { baseUrl?: string; apiKey?: string; authCallbacks?: DeviceFlowCallbacks }
): Promise<ModelInfo[]> {
  // For Copilot, we need to authenticate first with callbacks,
  // then create a provider with the resolved token
  if (type === "github-copilot") {
    const token = await getCopilotToken(opts?.authCallbacks);
    const provider = new OpenAICompatibleProvider({
      name: "github-copilot",
      model: "",
      baseUrl: opts?.baseUrl ?? "https://api.githubcopilot.com",
      apiKey: token,
    headers: { "Copilot-Integration-Id": "vscode-chat" },
    });
    return provider.listModels();
  }

  const config: ProviderConfig = {
    type,
    model: "",
    baseUrl: opts?.baseUrl,
    apiKey: opts?.apiKey,
  };
  const provider = createProvider(config);
  return provider.listModels();
}

File: rust/crates/core/src/providers/factory.rs
//! Provider presets and construction, porting
//! `packages/core/src/providers/factory.ts`.

use std::sync::Arc;

use crate::error::Result;
use crate::providers::LlmProvider;
use crate::providers::openai_compatible::{Auth, OpenAiCompatibleProvider, OpenAiProviderOptions};
use crate::types::{ModelInfo, ProviderConfig};

const COPILOT_INTEGRATION_HEADER: (&str, &str) = ("Copilot-Integration-Id", "vscode-chat");

/// Resolve a GitHub Copilot bearer token.
///
/// Indirection so the provider does not depend on the auth module's internals.
pub(crate) async fn copilot_bearer() -> Result<String> {
    crate::providers::github_auth::copilot_token(None).await
}

/// Build a provider from config.
///
/// Known types get a preset; anything else is treated as a raw
/// OpenAI-compatible endpoint, matching the TypeScript fallback.
pub fn create_provider(config: &ProviderConfig) -> Arc<dyn LlmProvider> {
    let name = config.label().to_string();

    let opts = match config.kind.as_str() {
        "ollama" => OpenAiProviderOptions {
            name,
            base_url: config
                .base_url
                .clone()
                .unwrap_or_else(|| "http://localhost:11434/v1".to_string()),
            auth: Auth::None,
            headers: Vec::new(),
        },

        "github-copilot" => OpenAiProviderOptions {
            name,
            base_url: config
                .base_url
                .clone()
                .unwrap_or_else(|| "https://api.githubcopilot.com".to_string()),
            auth: Auth::Copilot,
            headers: vec![(
                COPILOT_INTEGRATION_HEADER.0.to_string(),
                COPILOT_INTEGRATION_HEADER.1.to_string(),
            )],
        },

        "openai" => OpenAiProviderOptions {
            name,
            base_url: config
                .base_url
                .clone()
                .unwrap_or_else(|| "https://api.openai.com/v1".to_string()),
            auth: match config
                .api_key
                .clone()
                .or_else(|| std::env::var("OPENAI_API_KEY").ok())
            {
                Some(key) => Auth::Static(key),
                None => Auth::None,
            },
            headers: Vec::new(),
        },

        // Unknown type: a custom OpenAI-compatible endpoint.
        _ => OpenAiProviderOptions {
            name,
            base_url: config.base_url.clone().unwrap_or_default(),
            auth: match config.api_key.clone() {
                Some(key) => Auth::Static(key),
                None => Auth::None,
            },
            headers: Vec::new(),
        },
    };

    Arc::new(OpenAiCompatibleProvider::new(opts))
}

/// List models for a provider type without building a full agent.
///
/// For `github-copilot` this may trigger the device auth flow.
pub async fn list_models_for_provider(
    kind: &str,
    base_url: Option<&str>,
    api_key: Option<&str>,
) -> Result<Vec<ModelInfo>> {
    let config = ProviderConfig {
        kind: kind.to_string(),
        model: String::new(),
        base_url: base_url.map(str::to_string),
        api_key: api_key.map(str::to_string),
        name: None,
    };
    create_provider(&config).list_models().await
}

#[cfg(test)]
mod tests {
    use super::*;

    fn config(kind: &str) -> ProviderConfig {
        ProviderConfig {
            kind: kind.to_string(),
            model: "m".to_string(),
            base_url: None,
            api_key: None,
            name: None,
        }
    }

    #[test]
    fn presets_name_themselves_after_their_type() {
        for kind in ["ollama", "github-copilot", "openai", "something-custom"] {
            assert_eq!(create_provider(&config(kind)).name(), kind);
        }
    }

    #[test]
    fn an_explicit_name_overrides_the_type() {
        let mut c = config("openai");
        c.name = Some("work".to_string());
        assert_eq!(create_provider(&c).name(), "work");
    }
}
